server/src/main/java/org/elasticsearch/transport/InboundHandler.java

```diff
@@ -111,18 +120,29 @@ private void messageReceived(TcpChannel channel, InboundMessage message) throws
                 }
                 // ignore if its null, the service logs it
                 if (handler != null) {
-                    if (header.isError()) {
-                        handlerResponseError(streamInput, handler);
+                    final StreamInput streamInput;
+                    if (message.getContentLength() > 0) {
```

Comment:
I think I'd have just done this rather than introducing thread-locals here, and mixed-version clusters be (slightly) damned.

```suggestion
                    if (message.getContentLength() > 0 || header.getVersion().equals(Version.CURRENT) == false) {
```

What you've got looks ok as long as we don't refactor the deserialisation onto a different thread at some point in the future, but that's going to be a tricky bug to track down if ever we do so... :grin:

Reply:
Yea but that's the least of my worries in that case :D (we already make a million assumptions that assume we never escape the stream from the current thread with the way we release the paged bytes).
Plus, I tried moving the deserialization off the IO thread in https://github.com/elastic/elasticsearch/pull/50138 with pretty mixed results performance wise (it's the same reason we moved the serialization onto the IO loop, we just waste a ton of buffers if we deserialize later, which really hurts with 16k+ page sizes).

Reply:
> I think I'd have just done this rather than introducing thread-locals here, and mixed-version clusters be (slightly) damned.

Yea that's better, applied your suggestion :)